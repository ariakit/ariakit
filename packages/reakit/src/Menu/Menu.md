Menu allows you to display grouped navigation actions:

```jsx
import { Menu } from "reakit"

<Menu>
  <Menu.Item onClick={() => alert("Hello!")}>Item 1</Menu.Item>
  <Menu.Item>Item 2</Menu.Item>
  <Menu.Divider backgroundColor="#eee" />
  <Menu.Item>Item 3</Menu.Item>
  <Menu.Item disabled>Item 4 (disabled)</Menu.Item>
  <Menu.Item>Item 5</Menu.Item>
</Menu>
```

If you pass in a `horizontal` prop, the menu content will be vertically aligned:

```jsx
import { Menu } from "reakit"

<Menu horizontal>
  <Menu.Item onClick={() => alert("Hello!")}>Item 1</Menu.Item>
  <Menu.Item>Item 2</Menu.Item>
  <Menu.Divider backgroundColor="#eee" />
  <Menu.Item>Item 3</Menu.Item>
  <Menu.Item disabled>Item 4 (disabled)</Menu.Item>
  <Menu.Item>Item 5</Menu.Item>
</Menu>
```
