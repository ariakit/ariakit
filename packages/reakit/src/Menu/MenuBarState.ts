import * as React from "react";
import {
  SealedInitialState,
  useSealedState,
} from "reakit-utils/useSealedState";
import {
  CompositeState,
  CompositeActions,
  CompositeInitialState,
  useCompositeState,
} from "../Composite";

export type MenuBarState = CompositeState & {
  /**
   * Stores the values of radios and checkboxes within the menu.
   */
  unstable_values: Record<string, any>;
};

export type MenuBarActions = CompositeActions & {
  /**
   * Updates checkboxes and radios values within the menu.
   */
  unstable_setValue: (name: string, value?: any) => void;
};

export type MenuBarInitialState = CompositeInitialState &
  Partial<Pick<MenuBarState, "unstable_values">>;

export type MenuBarStateReturn = MenuBarState & MenuBarActions;

export function useMenuBarState(
  initialState: SealedInitialState<MenuBarInitialState> = {}
): MenuBarStateReturn {
  const {
    orientation = "horizontal",
    unstable_values: initialValues = {},
    ...sealed
  } = useSealedState(initialState);

  const [values, setValues] = React.useState(initialValues);
  const composite = useCompositeState({ ...sealed, orientation });

  return {
    ...composite,
    unstable_values: values,
    unstable_setValue: React.useCallback((name, value) => {
      setValues((vals) => ({
        ...vals,
        [name]: typeof value === "function" ? value(vals) : value,
      }));
    }, []),
  };
}
