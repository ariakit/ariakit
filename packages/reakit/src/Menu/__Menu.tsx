// TODO: Rename StaticMenu to Menu
import * as React from "react";
import { warning } from "reakit-utils/warning";
import { createComponent } from "reakit-system/createComponent";
import { useCreateElement } from "reakit-system/useCreateElement";
import { createHook } from "reakit-system/createHook";
import { usePipe } from "reakit-utils/usePipe";
import { mergeRefs } from "reakit-utils/mergeRefs";
import { BoxOptions, BoxHTMLProps, useBox } from "../Box/Box";
import { useShortcuts } from "./__utils/useShortcuts";
import { MenuContext } from "./__utils/MenuContext";
import { MenuStateReturn, useMenuState } from "./__MenuState";

export type StaticMenuOptions = BoxOptions &
  Pick<Partial<MenuStateReturn>, "orientation"> &
  Pick<MenuStateReturn, "stops" | "move" | "next" | "previous">;

export type StaticMenuHTMLProps = BoxHTMLProps;

export type StaticMenuProps = StaticMenuOptions & StaticMenuHTMLProps;

export const useStaticMenu = createHook<StaticMenuOptions, StaticMenuHTMLProps>(
  {
    name: "StaticMenu",
    compose: useBox,
    useState: useMenuState,

    useProps(options, { ref: htmlRef, unstable_wrap: htmlWrap, ...htmlProps }) {
      const ref = React.useRef<HTMLElement>(null);
      const parent = React.useContext(MenuContext);

      const providerValue = React.useMemo(
        () => ({
          orientation: options.orientation,
          next: options.next,
          previous: options.previous,
          parent
        }),
        [options.orientation, options.next, options.previous, parent]
      );

      useShortcuts(ref, options);

      const wrap = React.useCallback(
        (children: React.ReactNode) => (
          <MenuContext.Provider value={providerValue}>
            {children}
          </MenuContext.Provider>
        ),
        [providerValue]
      );

      return {
        ref: mergeRefs(ref, htmlRef),
        role: "menu",
        "aria-orientation": options.orientation,
        unstable_wrap: usePipe(wrap, htmlWrap),
        ...htmlProps
      };
    }
  }
);

export const StaticMenu = createComponent({
  as: "div",
  useHook: useStaticMenu,
  useCreateElement: (type, props, children) => {
    warning(
      !props["aria-label"] &&
        !props["aria-labelledby"] &&
        props.role !== "menubar",
      "StaticMenu",
      "You should provide either `aria-label` or `aria-labelledby` props.",
      "See https://reakit.io/docs/menu"
    );

    return useCreateElement(type, props, children);
  }
});
