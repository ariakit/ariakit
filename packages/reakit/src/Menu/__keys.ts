// Automatically generated
const MENU_BAR_STATE_KEYS = [
  "baseId",
  "unstable_idCountRef",
  "unstable_virtual",
  "rtl",
  "orientation",
  "items",
  "groups",
  "currentId",
  "loop",
  "wrap",
  "unstable_moves",
  "unstable_angular",
  "unstable_hasActiveWidget",
  "unstable_values",
  "setBaseId",
  "registerItem",
  "unregisterItem",
  "registerGroup",
  "unregisterGroup",
  "move",
  "next",
  "previous",
  "up",
  "down",
  "first",
  "last",
  "sort",
  "unstable_setVirtual",
  "setRTL",
  "setOrientation",
  "setCurrentId",
  "setLoop",
  "setWrap",
  "reset",
  "unstable_setHasActiveWidget",
  "unstable_setValue",
] as const;
const MENU_STATE_KEYS = [
  ...MENU_BAR_STATE_KEYS,
  "visible",
  "animated",
  "animating",
  "show",
  "hide",
  "toggle",
  "setVisible",
  "setAnimated",
  "stopAnimation",
  "modal",
  "unstable_disclosureRef",
  "setModal",
  "unstable_referenceRef",
  "unstable_popoverRef",
  "unstable_arrowRef",
  "unstable_popoverStyles",
  "unstable_arrowStyles",
  "unstable_originalPlacement",
  "unstable_update",
  "placement",
  "place",
] as const;
export const MENU_KEYS = MENU_STATE_KEYS;
export const MENU_ARROW_KEYS = MENU_KEYS;
export const MENU_BAR_KEYS = MENU_ARROW_KEYS;
export const MENU_BUTTON_KEYS = MENU_BAR_KEYS;
export const MENU_DISCLOSURE_KEYS = MENU_BUTTON_KEYS;
export const MENU_GROUP_KEYS = MENU_DISCLOSURE_KEYS;
export const MENU_ITEM_KEYS = MENU_GROUP_KEYS;
export const MENU_ITEM_CHECKBOX_KEYS = [...MENU_ITEM_KEYS, "name"] as const;
export const MENU_ITEM_RADIO_KEYS = MENU_ITEM_CHECKBOX_KEYS;
export const MENU_SEPARATOR_KEYS = MENU_ITEM_KEYS;
