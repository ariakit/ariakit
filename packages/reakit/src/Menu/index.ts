export * from "./Menu";
export * from "./MenuArrow";
export * from "./MenuBar";
export * from "./MenuBarState";
export * from "./MenuButton";
export * from "./MenuDisclosure";
export * from "./MenuGroup";
export * from "./MenuItem";
export * from "./MenuItemCheckbox";
export * from "./MenuItemRadio";
export * from "./MenuSeparator";
export * from "./MenuState";
