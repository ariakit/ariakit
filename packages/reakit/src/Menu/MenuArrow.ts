import { createComponent } from "reakit-system/createComponent";
import { createHook } from "reakit-system/createHook";
import {
  PopoverArrowOptions,
  PopoverArrowHTMLProps,
  usePopoverArrow,
} from "../Popover/PopoverArrow";
import { MENU_ARROW_KEYS } from "./__keys";

export type MenuArrowOptions = PopoverArrowOptions;

export type MenuArrowHTMLProps = PopoverArrowHTMLProps;

export type MenuArrowProps = MenuArrowOptions & MenuArrowHTMLProps;

export const useMenuArrow = createHook<MenuArrowOptions, MenuArrowHTMLProps>({
  name: "MenuArrow",
  compose: usePopoverArrow,
  keys: MENU_ARROW_KEYS,
});

export const MenuArrow = createComponent({
  as: "div",
  memo: true,
  useHook: useMenuArrow,
});
