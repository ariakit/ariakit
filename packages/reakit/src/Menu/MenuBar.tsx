import * as React from "react";
import { useWarning } from "reakit-warning";
import { createComponent } from "reakit-system/createComponent";
import { useCreateElement } from "reakit-system/useCreateElement";
import { createHook } from "reakit-system/createHook";
import { useForkRef } from "reakit-utils/useForkRef";
import {
  CompositeOptions,
  CompositeHTMLProps,
  useComposite,
} from "../Composite/Composite";
import { useShortcuts } from "./__utils/useShortcuts";
import { useMenuContext } from "./__utils/MenuContext";
import { MenuStateReturn } from "./MenuState";
import { MENU_BAR_KEYS } from "./__keys";

export type MenuBarOptions = CompositeOptions &
  Pick<Partial<MenuStateReturn>, "orientation"> &
  Pick<MenuStateReturn, "items" | "move" | "next" | "previous">;

export type MenuBarHTMLProps = CompositeHTMLProps;

export type MenuBarProps = MenuBarOptions & MenuBarHTMLProps;

export const useMenuBar = createHook<MenuBarOptions, MenuBarHTMLProps>({
  name: "MenuBar",
  compose: useComposite,
  keys: MENU_BAR_KEYS,

  useProps(
    options,
    {
      ref: htmlRef,
      wrapElement: htmlWrapElement,
      role = "menubar",
      ...htmlProps
    }
  ) {
    const ref = React.useRef<HTMLElement>(null);
    const wrap = useMenuContext(ref, role, options);

    useShortcuts(ref, options);

    const wrapElement = React.useCallback(
      (element: React.ReactNode) => {
        element = wrap(element);
        if (htmlWrapElement) {
          return htmlWrapElement(element);
        }
        return element;
      },
      [wrap, htmlWrapElement]
    );

    return {
      ref: useForkRef(ref, htmlRef),
      role,
      "aria-orientation": options.orientation,
      wrapElement,
      ...htmlProps,
    };
  },
});

export const MenuBar = createComponent({
  as: "div",
  useHook: useMenuBar,
  useCreateElement: (type, props, children) => {
    useWarning(
      !props["aria-label"] &&
        !props["aria-labelledby"] &&
        props.role !== "menubar",
      "You should provide either `aria-label` or `aria-labelledby` props.",
      "See https://reakit.io/docs/menu"
    );

    return useCreateElement(type, props, children);
  },
});
