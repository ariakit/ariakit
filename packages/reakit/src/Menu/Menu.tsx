import * as React from "react";
import { useWarning } from "reakit-warning";
import { createHook } from "reakit-system/createHook";
import { createComponent } from "reakit-system/createComponent";
import { useCreateElement } from "reakit-system/useCreateElement";
import { createOnKeyDown } from "reakit-utils/createOnKeyDown";
import { useLiveRef } from "reakit-utils/useLiveRef";
import { contains } from "reakit-utils/contains";
import {
  PopoverOptions,
  PopoverHTMLProps,
  usePopover,
} from "../Popover/Popover";
import { MenuBarOptions, MenuBarHTMLProps, useMenuBar } from "./MenuBar";
import { MenuStateReturn } from "./MenuState";
import { MenuContext, MenuContextType } from "./__utils/MenuContext";
import { MENU_KEYS } from "./__keys";

export type MenuOptions = Omit<PopoverOptions, "hideOnEsc"> &
  Pick<MenuStateReturn, "placement"> &
  MenuBarOptions;

export type MenuHTMLProps = PopoverHTMLProps & MenuBarHTMLProps;

export type MenuProps = MenuOptions & MenuHTMLProps;

function usePlacementDir(placement?: string) {
  return React.useMemo(() => placement?.split("-")?.[0], [placement]);
}

export const useMenu = createHook<MenuOptions, MenuHTMLProps>({
  name: "Menu",
  compose: [useMenuBar, usePopover],
  keys: MENU_KEYS,

  useOptions(options) {
    const parent = React.useContext(MenuContext);
    const parentIsMenuBar = parent?.role === "menubar";

    return {
      unstable_autoFocusOnHide: !parentIsMenuBar,
      modal: false,
      ...options,
      // will be handled by MenuButton
      unstable_autoFocusOnShow: false,
      // will be handled differently from usePopover/useDialog
      hideOnEsc: false,
    };
  },

  useProps(options, { onKeyDown: htmlOnKeyDown, ...htmlProps }) {
    const onKeyDownRef = useLiveRef(htmlOnKeyDown);
    const parent = React.useContext(MenuContext);
    const hasParent = !!parent;
    let ancestorMenuBar: MenuContextType | undefined | null = parent;

    while (ancestorMenuBar && ancestorMenuBar.role !== "menubar") {
      ancestorMenuBar = ancestorMenuBar.parent;
    }

    const { next, previous, orientation } = ancestorMenuBar || {};
    const ancestorIsHorizontal = orientation === "horizontal";
    const dir = usePlacementDir(options.placement);

    const onKeyDown = React.useMemo(
      () =>
        createOnKeyDown({
          onKeyDown: onKeyDownRef,
          stopPropagation: (event) => {
            // On Esc, only stop propagation if there's no parent menu.
            // Otherwise, pressing Esc should close all menus
            if (hasParent) {
              return event.key !== "Escape";
            }
            return event.key === "Escape";
          },
          keyMap: ({ currentTarget, target }) => {
            const { hide } = options;
            const close = hide && (() => hide());
            if (hasParent && contains(currentTarget, target as Element)) {
              // Moves to the next menu button in a horizontal menu bar or
              // close the menu if it's a sub menu
              const ArrowRight =
                ancestorIsHorizontal && dir !== "left"
                  ? next && (() => next())
                  : dir === "left" && close;
              const ArrowLeft =
                ancestorIsHorizontal && dir !== "right"
                  ? previous && (() => previous())
                  : dir === "right" && close;
              return { Escape: close, ArrowRight, ArrowLeft };
            }
            return { Escape: close };
          },
        }),
      [hasParent, ancestorIsHorizontal, next, previous, dir, options.hide]
    );

    return { role: "menu", onKeyDown, ...htmlProps };
  },
});

export const Menu = createComponent({
  as: "div",
  useHook: useMenu,
  useCreateElement: (type, props, children) => {
    useWarning(
      !props["aria-label"] && !props["aria-labelledby"],
      "You should provide either `aria-label` or `aria-labelledby` props.",
      "See https://reakit.io/docs/menu"
    );
    return useCreateElement(type, props, children);
  },
});
