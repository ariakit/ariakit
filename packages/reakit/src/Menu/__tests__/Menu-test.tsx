import * as React from "react";
import { render } from "reakit-test-utils";
import { Menu } from "../Menu";

function createRef(id: string) {
  const ref = React.createRef() as React.MutableRefObject<HTMLElement>;
  ref.current = document.createElement("div");
  ref.current.id = id;
  return ref;
}

const props: Parameters<typeof Menu>[0] = {
  baseId: "base",
  items: [
    { id: "a", ref: createRef("a") },
    { id: "b", ref: createRef("b") },
  ],
  setCurrentId: jest.fn(),
  move: jest.fn(),
  next: jest.fn(),
  previous: jest.fn(),
  first: jest.fn(),
  last: jest.fn(),
  placement: "bottom-start",
  "aria-label": "menu",
};

test("render", () => {
  const { baseElement } = render(<Menu {...props} />);
  expect(baseElement).toMatchInlineSnapshot(`
    <body>
      <div>
        <div
          aria-label="menu"
          data-dialog="true"
          hidden=""
          id="base"
          role="menu"
          style="display: none;"
          tabindex="-1"
        />
      </div>
    </body>
  `);
});

test("render without state props", () => {
  // @ts-ignore
  const { baseElement } = render(<Menu id="base" aria-label="menu" />);
  expect(baseElement).toMatchInlineSnapshot(`
    <body>
      <div>
        <div
          aria-label="menu"
          data-dialog="true"
          hidden=""
          id="base"
          role="menu"
          style="display: none;"
          tabindex="-1"
        />
      </div>
    </body>
  `);
});
