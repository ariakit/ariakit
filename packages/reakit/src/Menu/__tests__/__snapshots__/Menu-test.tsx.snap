// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0[aria-orientation="horizontal"] {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

<ul
  aria-orientation="vertical"
  class="c0 c1"
  role="menu"
/>
`;

exports[`styled horizontal 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0[aria-orientation="horizontal"] {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

<ul
  aria-orientation="horizontal"
  class="c0 c1"
  role="menu"
/>
`;
